core/libs/models/src/tree.rs

```diff
@@ -123,133 +164,122 @@ where
+                    Some((*id, file.clone()))
+                } else {
+                    None
+                }
+            })
             .collect()
     }
 
-    /// Returns the files which are deleted or have deleted ancestors. It is an error for the parent
-    /// of a file argument not to also be included in the arguments.
-    fn filter_deleted(&self) -> Result<Vec<Fm>, TreeError> {
-        let mut result = Vec::new();
-        for file in self {
+    fn filter_deleted(&self) -> Result<HashMap<Uuid, Fm>, TreeError> {
+        // todo!() //look into optimizing. Optimized: I think it's O(n) instead of O(n^2) now
+        let mut result = HashMap::new();
+        let mut not_deleted = HashMap::new();
+        for (id, file) in self {
+            let mut ancestors = HashMap::from([(*id, file.clone())]);
             let mut ancestor = file.clone();
             loop {
-                if ancestor.deleted() {
-                    result.push(file.clone());
+                if not_deleted.get(&ancestor.id()).is_none() // check it isn't confirmed as not deleted
+                    && (ancestor.deleted() || result.get(&ancestor.id()).is_some())
+                {
+                    result.extend(ancestors);
                     break;
                 }
 
                 let parent = self.find(ancestor.parent())?;
-                if ancestor.id() == parent.id() {
-                    break;
+                // first case is root, second case is a cycle (not our problem)
+                if parent.id() == ancestor.id() || &parent.id() == id {
+                    not_deleted.extend(ancestors);
+                    break; // root
                 }
+                ancestors.insert(parent.id(), parent.clone());
                 ancestor = parent;
-                if ancestor.id() == file.id() {
-                    break; // this is a cycle but not our problem
-                }
             }
         }
         Ok(result)
     }
 
-    /// Returns the files which are not deleted and have no deleted ancestors. It is an error for
-    /// the parent of a file argument not to also be included in the arguments.
-    fn filter_not_deleted(&self) -> Result<Vec<Fm>, TreeError> {
+    fn filter_not_deleted(&self) -> Result<HashMap<Uuid, Fm>, TreeError> {
         let deleted = self.filter_deleted()?;
         Ok(self
-            .iter()
-            .filter(|f| !deleted.iter().any(|nd| nd.id() == f.id()))
-            .cloned()
+            .clone()
+            .into_iter()
+            .filter(|(id, _)| deleted.get(id).is_none())
             .collect())
     }
 
-    /// Returns the files which are documents.
-    fn filter_documents(&self) -> Vec<Fm> {
-        self.iter()
-            .filter(|f| f.file_type() == FileType::Document)
-            .cloned()
+    fn filter_documents(&self) -> HashMap<Uuid, Fm> {
+        self.clone()
+            .into_iter()
+            .filter(|(_, f)| f.file_type() == FileType::Document)
```

Comment:
```suggestion
            .filter(|(_, f)| f.file_type() == Document)
```
`FileType`'s variants got imported for this file with this PR. So if we are going to keep the variants, then let's drop the enum prefix just right here for now.